### DISCUSSION TITLE
Sharding application-controller by Application or a Project

### DISCUSSION THREAD

**Discussion by msobkowiak-olx** [argoproj/argo-cd#14346] [General]
Hey there !

Wanted to ask on possibilities to distribute load across multiple application-controllers, our setups usually consist of one ArgoCD instance per cluster, thus is not very beneficial for us to use default sharding option described here:
https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/

I would like to ask if there is any plans to actually change this to shard per application name (using static or consistent (with crash support) hashing, just like on memcached clients or in loadbalancers), which should have the following benefits:
- best possible granularity of the load (with monorepos still being sort of an issue however it is still an issue even now)
- best resource ultilisation (achieving best possible "even" distribution, with exception for large monorepos)
- best possible scaling scenarios (supporting ArgoCD with one or multiple clusters)

Similar can be applied to a project. The hashing can simply take project or app name, calculate the hash and send it to proper slot for processing in the deployed statefulset. This should be possible because of the constraints ArgoCD puts on the Applications and Projects (unique naming, guaranteeing that each entity has unique key for hashing function, the distribution happens onto application-controller pods).

Please note I am aware of the queue tuning on operation and status processors, however we would like to avoid creating too large pods not fitting our workload nodes in the clusters hosting ArgoCD.

Let me know if it makes sense.

**Comment by awx-fuyuanchu** [comment]
Yeah. It sounds like a more general pain point for most users. One cluster could spawn hundreds of Applications and sharding the Applications in one cluster can help a lot for performance, efficiency, and scalability.

**Comment by saprette** [comment]
Hello, that is something that we would benefit from.

**Comment by andrii-korotkov-verkada** [comment]
I suggest to open an enhancement proposal for this. I don't know if this is easily solvable, since resources can move between applications and applications can move between projects.

**Reply by msobkowiak-olx** [reply]
For application controller I do not think it really matters where the application is placed or what the resources are inside of it. The purpose of application controller is to grab the application and process its state (so run the entire sync/reconciliation thing). That is why these components are still "kinda stateless" (except they have to know total number of nodes and their ID which is a simple problem to solve with the ENV vars and the statefulset deployment).

It is really all about how application controller chooses the entities to process (**aka assigning applications to shards**). I assume that cluster list is perhaps something that rarely changes and this was a primary argument when designing sharding in general (reshuffling the clusters among the appcontrollers or, in other words, assigning clusters to shards is much less dynamic and is much more predictable and easier to control). On the other hand Application List can change very often and it needs to be re-discovered and assigned to a controller somehow.

While perhaps sharding by project is for sure susceptible to problems, application is really the smallest granulation unit in ArgoCD so it offers the best sharding distribution anyhow.

**Reply by andrii-korotkov-verkada** [reply]
What if some resource changes applications. Which shard would process that?

**Reply by msobkowiak-olx** [reply]
Nothing should change in this case really, both applications will be processed independently by their respective shards achieving eventual consistency and the state will eventually be reflected properly (the resource will disappear from app A and will appear in App B).

**Reply by andrii-korotkov-verkada** [reply]
I think the problem is two things can maintain a different view of the same object in Kubernetes, so there might be conflicting changes.

**Reply by msobkowiak-olx** [reply]
Isn't this an issue even now on single shard ? I mean you can technically deploy two different applications with different name and exactly the same manifests (we used to hit collisions sometimes during migrations from helm).

**Reply by andrii-korotkov-verkada** [reply]
There's a cluster lock acquired by a number of places.

**Reply by msobkowiak-olx** [reply]
Thanks, I mean it would be crucial to understand what exactly this lock synchronizes in this case. Implementing a distributed lock would not be that easy (technically the locking for cross-shard implementation could be externalised to redis or so - that is actually not that hard but prone to process crashes without expiration implementation).

**Reply by andrii-korotkov-verkada** [reply]
I can bring this up during the contributors meeting.

**Comment by andrii-korotkov-verkada** [comment]
See https://github.com/argoproj/argo-cd/issues/6125 for a similar thing.

**Comment by crenshaw-dev** [comment]
Most of the work that the Application controller does is cluster-scoped. Breaking it into Application- or Project-sized chunks is very high technical cost for potentially very low reward.

In my opinion, the time is better spent improving the general performance of the Application controller and tuning performance configuration to alleviate issues.

**Reply by stormmore** [reply]
It feels like you are making an assumption that people with multiple cluster run a single argocd architecture.

As someone who runs single argocd per cluster, I would argue that Application/Project chunks is very high reward, especially if I can shard critical application out from not so critical. The only sharding method we have that gives us any granularity is to run multiple argocds - one for each grouping - which is basically what our platform vendor recommends as well.

For the record, one of the clusters that I am an engineer of currently uses ArgoCD to manage 180 "applications" and we have just pushed its single controller up to 1.5 core / 6 GB request and 4 core / 12 GB due to OOMKills and slow sync speeds.

**Reply by crenshaw-dev** [reply]
If the main motivation is to mitigate "noisy neighbor" issues, I think a prod/pre-prod Argo CD instance split makes a lot of sense and is probably a better idea than having that split within just the application-controller. Presumably you'd also want to protect the repo-server, applicationset-controller, Dex, etc. from noisy pre-prod neighbors.

**Reply by luispabon** [reply]
Our use case is one cluster, one argocd instance managing hundreds of applications in that cluster. We've been forced to scale the application controller's node to very high levels (16 vCPUs per node)  just to cope with the load created on a single pod.

Redis being available, can't it not be used as ephemeral state to coordinate which controller replica is working on what? Not knowing argo's internal design I can think of no reason this particular job can't be pub/subbed or some other similar mechanism for assigning work.

**Reply by chris-ng-1987** [reply]
Agree with @luispabon, I feel bad as ArgoCD is a CNCF project, but the workload of application controller is not supporting distribute single cluster applications workload by horizontal scaling, forcing vertical scaling without HA.

**Reply by 7onn** [reply]
In a staging cluster, besides the apps themselves, I also provision preview environments where in each namespace, I have to render "my whole company"'s manifests; It is a lot to handle single-shardedly; and ArgoCD became really slow; sometimes taking nearly a minute to get ready to sync changes that people push on their pull requests.

It was a sad suprise when I enabled sharding just to find out all manifests were still falling into the same shard, leaving other two instances completely idle.

My current scale in the staging cluster is ~125 applications; but the preview ones (~90) can be very heavy, carrying from 26 to 80 resource's yaml per application! They are provisioned by pull request, depending on which apps in our monorepo are changing on a given PR.

So I can say it would be extremely rewarding for us if I could balance these apps' manifests across more ArgoCD app controller instances;

**Reply by icy** [reply]
@7onn @msobkowiak-olx

I have found a work-around. By checking my k8s control api endpoint I can see they normally support 4 DNS and 1 private ip.

```
$ echo |openssl s_client -connect kubernetes.default:443|openssl x509 -noout -text |grep "Alternative Name"

X509v3 Subject Alternative Name: 
                DNS:kubernetes, DNS:kubernetes.default, DNS:kubernetes.default.svc, 
               DNS:kubernetes.default.svc.cluster.local, IP Address: 172.173.0.9
```

By default, argocd picks up `kubernetes.default.svc` for its `in-cluster` configuration. That means, in practice, you can configure another 4 server endpoint `kubernetes`, `kubenetest.default`, `kubernetes.default.svc.cluster.local` and `172.173.0.9`. 

I've  tested successfully with this idea and put 50% of my applications into the second replica of the application controller. My setup has a single argocd server [that] runs within and manages all resources of a k8s cluster (cluster-scope).

I hope this helps.

**Reply by msobkowiak-olx** [reply]
Thanks, I was thinking about the same, allegedly there are still potential locking/caching issues that can occur with this as stated above:

> There's a cluster lock acquired by a number of places.

I just don't know if this is by cluster "name" or if it uses a centralized dependency. If this is just a name in redis/controller memory it should work absolutely fine.

If this approach works then this problem can be easily solved by introducing cluster aliasing.

**Reply by icy** [reply]
@msobkowiak-olx I understood, thanks for your sharing. My setup has about 700 (ArgoCd) applications on a single k8s cluster. Fortunately they are quite independently from each other, so I don't need to worry about cache or alike. I haven't noticed any issue.

**Reply by 7onn** [reply]
@icy thanks a lot for sharing, bro! That's definitely helpful as now I can have a separate ArgoCD project for my previews (pointing to the "same" cluster) and split their load from the rest of my stable apps.

But that still puts the vast majority of my apps in the same shard; So I guess sharding by app would still be valuable for a fair and balanced workload distribution.

**Reply by icy** [reply]
@7onn totally agreed; the  work-around is very limited and manually. Better sharding strategy is really expected.

**Reply by msobkowiak-olx** [reply]
If the applications are consistent (with 700 you would for sure hit those issues) I think the purpose of the locking is to simply have the application refreshed by exactly one application controller at the time. If this is the case honestly I think sharding and locks per app should be possible. Perhaps problematic part could be resharding/adding apps (as this is something that changes often) but I would say the ring can be just composed by application controller names (statefulset would really help here to keep the ring stable) and then consistent hashing could be done by application name.

I guess what the issue is here is that each applicationcontroller on startup needs to have a list of apps assigned as there is no "coordinator" concept here (that is usually an issue with operators pattern - they elect a leader for given partition and here its definetely easier to partition by cluster so we dont re-discover every now and then). To solve that the stateless coordinator concept could be introduced to hash/balance the traffic and pass the requests to applicationcontrollers given by hashing/balancing algorithm (just like in regular loadbalancers) but this then needs to externalize the locks (eg with centralized redis so no concurrent changes are made to the single app with multiple applicationcontroller at the same time).

Again - lazy solution could be allowing to add the cluster with multiple aliases - that point to the same host but different name (ARNs or other unique metadata would then be a bit tricky to use tho).

**Reply by jessesuen** [reply]
The technical challenge is not just about spreading work around. The challenge is that, to properly reconcile an application, the app-controller needs to understand the full state of the cluster, to maintain what we call the "live state cache." The live state cache is an in-memory cache that stores lightweight metadata (name, namespace, gvk, ownerResources) about all resources, which is then used to:
1. detect when a resource (e.g. Pod) is related to an application so that an application can be queued for reconcilliation
2. reconstruct the application tree during reconciliation

Even if we were to support multiple shards for a single cluster, each shard would still have to maintain its own live state cache, and each shard would use the same amount of memory resources, **_even if_** they were handling only a subset of the apps for that cluster. Not to mention, since each live state cache would be independent of the others, there could be inconsistencies between them, which then implies this live state cache would need be split out as a separate component for consistency. Accessing this cache then becomes an RPC call rather than in-memory access, slowing reconciliation.

This has been considered before, but thus far, nothing I've heard can easily overcome the challenges I've described above.

**Reply by 7onn** [reply]
@icy how did you manage to create the same cluster twice? I couln't succeed on this as a new secret requires me to [input caData and certData](https://argo-cd.readthedocs.io/en/stable/operator-manual/declarative-setup/#clusters); while the default `in-cluster` seems to manage this by itself.

I fear that creating them with these hardcoded values would make me vulnerable as it would break my implementation as soon as EKS rotates the certificate. Plus, the config doesn't allow me to reference certs from a file so I could use the one ArgoCD's service account mounts in the pods.

Did you toggle the `tlsClientConfig.insecure: true`?

**Reply by icy** [reply]
Hi @7onn ,

Sorry I didn't mention that before to avoid some noise. Here are all things that helped in my situation. I hope that helps. 

I use `kubectl` to enter any  argocd (application controller) pod, from within the pod, I can try some basic `kubectl` commands for example `kubectl get pods`, ` kubectl get nodes`; these commands work in my setup. Then I create a new `~/.kube/config` file (as below). You may add as many configurations as you need, just keep in mind that `server:` values should be unique. The next step is very trivial

```
$ argocd login
$ argocd cluster add clone1
$ argocd cluster add clone2
```

`argocd` will warn if you want to overwrite some existing cluster. Be careful to answer ;) In my case, I just run with `--upsert` option and overwrite everything.

## Authentication

You may notice the configuration relies on the default service token `/var/run/secrets/kubernetes.io/serviceaccount/token`. This has some expiration, 6 month or one year (depends on your setup). But that doesn't really matter. I just need to re-run the command `argocd cluster add clone1` as above every now and then.

## My `~/.kube/config`

NOTE: this file is created with the argocd application controller pod.

```
apiVersion: v1
kind: Config
clusters:
- name: clone1
  cluster:
    server: https://kubernetes.default
    certificate-authority: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
- name: clone2
  cluster:
    server: https://kubernetes.default.svc.cluster.local
    certificate-authority: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
users:
- name: clone1-sa
  user:
    tokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
- name: clone2-sa
  user:
    tokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
contexts:
- name: clone1
  context:
    cluster: clone1
    user: clone1-sa
    namespace: default
- name: clone2
  context:
    cluster: clone2
    user: clone2-sa
    namespace: default
current-context: clone1
```

**Reply by icy** [reply]
PS: Please also note I removed the additional clusters (clone2, clone3,...) in the sample. It's just a lot of duplication.
